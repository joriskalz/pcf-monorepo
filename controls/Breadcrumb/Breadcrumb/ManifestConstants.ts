export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Clickable = 'ItemClickable',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
