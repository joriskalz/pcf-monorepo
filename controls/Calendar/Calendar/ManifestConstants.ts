export const enum InputEvents {
    SetFocus = 'SetFocus',
}
