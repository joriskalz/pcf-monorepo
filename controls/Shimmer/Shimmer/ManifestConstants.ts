export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    Key = 'ItemKey',
    Type = 'ItemType',
    Height = 'ItemHeight',
    Width = 'ItemWidth',
    VerticalAlign = 'ItemVerticalAlign',
    RowKey = 'ItemRowKey',
}

export const enum RowColumns {
    Key = 'RowKey',
    Order = 'RowOrder',
    Count = 'RowCount',
    Width = 'RowWidth',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
    Reset = 'Reset',
}

export const enum InputProperties {
    SelectedKey = 'SelectedKey',
}

export class RenderTypes {
    static PivotLinks = '0';
    static PivotTabs = '1';
}

export class RenderSize {
    static Normal = '0';
    static Compact = '1';
    static Large = '2';
}
