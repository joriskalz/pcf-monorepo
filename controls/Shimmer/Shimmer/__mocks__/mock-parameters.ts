/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        Theme: new MockStringProperty(),
        AccessibilityLabel: new MockStringProperty(),
        RowCount: new MockWholeNumberProperty(),
        items: new MockDataSet([]),
        SpacebetweenShimmer: new MockEnumProperty(),
    };
}
