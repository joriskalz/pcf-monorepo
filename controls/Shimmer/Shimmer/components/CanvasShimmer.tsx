import * as React from 'react';
import { Shimmer, ThemeProvider, createTheme, IPartialTheme, mergeStyles } from '@fluentui/react';
import { IShimmerProps } from './Component.types';

export const CanvasShimmer = React.memo((props: IShimmerProps) => {
    const { shimmerElements, themeJSON, spacebetweenShimmer, rowCount } = props;
    const wrapperClass = mergeStyles({
        padding: 2,
        selectors: {
            '& > .ms-Shimmer-container': {
                margin: spacebetweenShimmer?.concat(' 0'),
            },
        },
        width:props.width
    });

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    return (
        <ThemeProvider theme={theme} className={wrapperClass}>
            {[...Array(rowCount)].map((key) => {
                return <Shimmer key={key} shimmerElements={shimmerElements} />;
            })}
        </ThemeProvider>
    );
});
CanvasShimmer.displayName = 'CanvasShimmer';
