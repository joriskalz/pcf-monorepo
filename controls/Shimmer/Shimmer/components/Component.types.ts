import { IShimmerElement } from '@fluentui/react';

export interface ICustomShimmerItem {
    id: string;
    type: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
    height: number;
    width: string;
    verticalAlign: string;
    rowkey: string;
}

export interface IShimmerRows {
    id: string;
    key: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    order: number;
    count: number;
    width: number;
}

export interface IShimmerProps {
    width?: number;
    height?: number;
    shimmerElements: IShimmerElement[];
    themeJSON?: string;
    spacebetweenShimmer?: string;
    rowCount?: number;
}
