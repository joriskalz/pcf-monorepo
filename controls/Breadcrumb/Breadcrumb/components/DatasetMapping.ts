import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
