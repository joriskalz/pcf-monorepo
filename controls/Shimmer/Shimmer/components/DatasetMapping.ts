import { ICustomShimmerItem } from './Component.types';
import { ItemColumns } from '../ManifestConstants';
import { IShimmerElement, ShimmerElementType } from '@fluentui/react';
export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomShimmerItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return getDummyItems();
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            type: record.getValue(ItemColumns.Type) as string,
            height: record.getValue(ItemColumns.Height) as number,
            width: record.getValue(ItemColumns.Width) as string,
            verticalAlign: record.getValue(ItemColumns.VerticalAlign) as string,
            data: record,
        } as ICustomShimmerItem;
    });
}

export function getShimmerElements(items: ICustomShimmerItem[]): IShimmerElement[] {
    return items.map((item: ICustomShimmerItem) => ({
        type: getShimmerElementType(item.type),
        ...(typeof item.width !== 'undefined' && item.width !== null && { width: item.width + '%' }),
        ...(typeof item.height !== 'undefined' && item.height !== null && { height: item.height }),
        ...(typeof item.verticalAlign !== 'undefined' && item.verticalAlign !== null && { width: item.verticalAlign }),
    })) as IShimmerElement[];
}

export function getShimmerElementType(shimmerElementType: string): ShimmerElementType {
    let currentElementType: ShimmerElementType = ShimmerElementType.line;
    switch (shimmerElementType.toLowerCase()) {
        case 'circle':
            currentElementType = ShimmerElementType.circle;
            break;
        case 'gap':
            currentElementType = ShimmerElementType.gap;
            break;
        default:
            currentElementType;
    }
    return currentElementType;
}

function getDummyItems(): ICustomShimmerItem[] {
    return [
        {
            width: '3.8',
            height: 25,
            type: 'circle',
        },
        {
            width: '5',
            height: 10,
            type: 'gap',
        },
        {
            width: '100',
            height: 20,
            type: 'line',
        },
        {
            width: '10',
            height: 10,
            type: 'gap',
        },
        {
            width: '100',
            height: 20,
            type: 'line',
        }
    ] as unknown as ICustomShimmerItem[];
}
