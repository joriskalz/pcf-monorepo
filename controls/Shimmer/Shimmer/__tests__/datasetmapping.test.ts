import { getItemsFromDataset, getShimmerElements, getShimmerElementType } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';
import { ShimmerElementType } from '@fluentui/react';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: '1',
                [ItemColumns.Height]: 10,
                [ItemColumns.Width]: '20',
                [ItemColumns.RowKey]: '1',
                [ItemColumns.Type]: 'gap',
            }),
        ];
        expect(getShimmerElementType('Circle')).toEqual(ShimmerElementType.circle);
        const shimmerItems = getShimmerElements(getItemsFromDataset(new MockDataSet(items)));
        expect(shimmerItems).toMatchSnapshot();
    });
});
