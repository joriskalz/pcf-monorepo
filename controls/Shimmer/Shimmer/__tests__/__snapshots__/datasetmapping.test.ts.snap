// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "height": 10,
    "type": 3,
    "width": "20%",
  },
]
`;
